src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs

```diff
@@ -897,6 +897,122 @@ public void SendRecvIovMax_Success()
+                }
+
+                SocketError error;
+                // send data as segmentCount (> IOV_MAX) 1-byte segments.
+                int bytesSent = socket.Send(sendSegments, SocketFlags.None, out error);
+                if (error == SocketError.Success)
+                {
+                    // platform sent message with > IOV_MAX segments
+                    Assert.Equal(segmentCount, bytesSent);
+                }
+                else
+                {
+                    // platform returns EMSGSIZE
+                    Assert.Equal(SocketError.MessageSize, error);
+                    Assert.Equal(0, bytesSent);
+                }
+            }
+        }
+
+        [Fact]
+        public async Task ReceiveIovMaxUdp_SuccessOrMessageSize()
+        {
+            // receiving more than IOV_MAX segments causes EMSGSIZE on some platforms.
+            // We handle this for stream sockets by truncating.
+            // This test verifies we are not truncating non-stream sockets.
+
+            // Use more than IOV_MAX (1024 on Linux & macOS) segments
+            // and less than Ethernet MTU.
+            const int segmentCount = 1200;
+            var receiver = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp);
+            receiver.BindToAnonymousPort(IPAddress.Loopback);
+            Task receiveTask = Task.Run(() =>
+            {
+                using (receiver)
+                {
+                    var receiveBuffer = new byte[segmentCount];
+                    var receiveSegments = new List<ArraySegment<byte>>();
+                    for (int i = 0; i < segmentCount; i++)
+                    {
+                        receiveSegments.Add(new ArraySegment<byte>(receiveBuffer, i, 1));
+                    }
+                    // receive data as segmentCount (> IOV_MAX) 1-byte segments.
+                    EventWaitHandle completed = new ManualResetEvent(false);
+                    var socketEvent = new SocketAsyncEventArgs();
+                    socketEvent.UserToken = completed;
+                    socketEvent.BufferList = receiveSegments;
+                    socketEvent.Completed += OnCompleted;
+                    if (receiver.ReceiveAsync(socketEvent))
+                    {
+                        completed.WaitOne();
+                    }
+
+                    SocketError error = socketEvent.SocketError;
+                    int bytesReceived = socketEvent.BytesTransferred;
+                    if (error == SocketError.Success)
+                    {
+                        // Linux sometimes returns EMSGSIZE and sometimes fills a single segment and sets MSG_TRUNC.
+                        // The latter is unexpected, MSG_TRUNC should only be set when there is insufficient buffer space.
+                        bool truncated = (socketEvent.SocketFlags & SocketFlags.Truncated) != SocketFlags.None;
```

Comment:
@stephentoub I've added tests. This test had some unexpected behavior on Linux as described in the comment.

Sometimes strace was expected EMSGSIZE:
```
[pid  5361] sendmsg(105, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"..., iov_len=1200}], msg_iovlen=1, msg_controllen=0, msg_flags=0}, 0) = 1200
[pid  5378] recvmsg(104, {msg_namelen=0}, 0) = -1 EMSGSIZE (Message too long)
```
sometimes receive filled 1 segment and set MSG_TRUNC:
```
[pid  5294] sendmsg(105, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"..., iov_len=1200}], msg_iovlen=1, msg_controllen=0, msg_flags=0}, 0) = 1200
[pid  5309] recvmsg(104, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="\0", iov_len=1}], msg_iovlen=1, msg_controllen=0, msg_flags=MSG_TRUNC}, 0) = 1
```

Reply:
@tmds, we believe this is an OS bug?

Reply:
I'm unable to reproduce it today :/ I'm going to remove this special case and see what CI makes of it.

Reply:
CI runs fine without it. Not sure what caused it for me yesterday, but I'm happy we can do without.